Modules tutorial (#27)
COVESA/ramses-composer-docs: advanced/modules/README.md

<!--
SPDX-License-Identifier: MPL-2.0

This file is part of Ramses Composer
(see https://github.com/COVESA/ramses-composer-docs).

This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
-->

# Lua Modules

![](../../basics/monkey/docs/viewport_preview.png)

> This tutorial demonstrates how to re-use Lua functions and configuration data using Lua modules.

Reviewer: ```suggestion
This tutorial demonstrates how to reuse Lua functions and configuration data using Lua modules.
```

## What are modules good for?

Any large project needs to reuse code or data at some point. This holds true for assets, in particular for Lua source code.
It's quite easy to reuse data/resources in the Ramses Composer - you just have to reference the same external file (png, glsl, glTF etc.).
Naturally, you can also have two script instances referencing the same Lua source file. The Ramses Composer will however still
create two *instances* of Lua scripts, which will still have their own state and data. The only thing they share is that both scripts
will be recreated if the source code changes. Moreover, if you want to slightly modify one of the instances, you have to create two mostly
identical Lua source files, and thus create code duplication.

Modules provide a solution to this problem which works very similar to how typical libraries or modules in other languages work.
You can put some of the functions and data you need in your project in a module, and reference that module from multiple scripts.
Let's build an example which does that!

> ## Copy pasta

Reviewer 2: ```suggestion
## Copy paste
```

To create a new module, we need to create a new Lua source file and write its code there. We follow the
[official Logic Engine docs](https://ramses-logic.readthedocs.io/en/latest/lua_syntax.html#custom-modules) and
create the following [Lua module source file](./scripts/LightModule.lua). You may notice that the file looks quite
similar to the LightControl script from the basic monkey example, with a few structural differences:

* The module has no global functions such as interface()
* Instead, it declares a table (called light) in the beginning of the file and returns it to the caller at the end
* In the table, the module puts data (color and light properties) and functions to obtain those (getLightDirection(), resolveColor())
* It has more error checks than the original LightControl script

We will explain later why this is necessary. Let's move on and use the new module first!

## Using the module

Modules have to be explicitly created and imported in the scripts which need to access them. To create a module,
right click the resources area and create a new object of type 'LuaScriptModule'. Set its URI to the source code file
we created in the previous section.

> To let a script make use of a module, you have to explicitly declare you want to use it. Add the following line in the

Reviewer 2: ```suggestion
To let a script make use of a module, you have to explicitly declare how you want to use it. Add the following line in the
```

LightControl script:

```lua
modules('light')

function interface()
...

```

> Now the LightControl script will report errors, because you declared a module but didn't provide one to the script. To fix the

Reviewer 2: What do you mean by "didn't provide one to the script"?

> Now the LightControl script will report errors, because you declared a module but didn't provide one to the script. To fix the

Reviewer: ```suggestion
Now the LightControl script will report errors, because you declared a module in the script file but didn't provide a module to the respective LuaScript object in Ramses Composer. To fix the
```

We included the new module as a dependency, but we are not using it yet. To use it, we can refer to all functions and data
members of the module using a table with the same name as declared in the ``modules()`` statement:

```lua

modules("light")

function interface()
   ...
end

function run()
    OUT.light_direction = light.getLightDirection(IN.light_id)
    OUT.light_color     = light.resolveColor("light", IN.light_color)
    OUT.diffuse_color   = light.resolveColor("diffuse", IN.diffuse_color)
end
```

Now, with this change, the LightControl Lua script delegates the configuration of the light properties
to the LightModule module. You can now add a second script, which uses the same basic configuration,
but modifies or extends it. For example, it could make the colors brigther, or it could substitute the
hardcoded diffuse color with a dynamic color which is animated. Feel free to experiment!

## Notes on the LightModule

The example module we created for this example has a slightly different syntax than other scripts in the Logic engine.
The first notable difference is that it declares no global functions or data, but instead declares a local table (the module),

> fills it with content, and returns it. The reason for that is that this is how Lua recommends to define modules

Reviewer: ```suggestion
fills it with content, and returns it. That's because this is how Lua recommends to define modules
```
